pkg/engine/lifecycletest/target_test.go

```diff
@@ -1599,3 +1600,356 @@ func TestTargetUntargetedParent(t *testing.T) {
 		assert.Equal(t, tokens.Type("pulumi:providers:pkgA"), snap.Resources[1].URN.Type())
 	})
 }
+
+// TestTargetDestroyDependencyErrors ensures we get an error when doing a targeted destroy of a resource that has a
+// dependency and the dependency isn't specified as a target and TargetDependents isn't set.
+func TestTargetDestroyDependencyErrors(t *testing.T) {
+	t.Parallel()
+
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				CreateF: func(urn resource.URN, news resource.PropertyMap, timeout float64,
+					preview bool,
+				) (resource.ID, resource.PropertyMap, resource.Status, error) {
+					return "created-id", news, resource.StatusOK, nil
+				},
+			}, nil
+		}, deploytest.WithoutGrpc),
+	}
+
+	programF := deploytest.NewLanguageRuntimeF(func(info plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		aURN, _, _, err := monitor.RegisterResource("pkgA:m:typA", "resA", true)
+		assert.NoError(t, err)
+
+		_, _, _, err = monitor.RegisterResource("pkgA:m:typA", "resB", true, deploytest.ResourceOptions{
+			Dependencies: []resource.URN{aURN},
+		})
+		assert.NoError(t, err)
+		return nil
+	})
+	hostF := deploytest.NewPluginHostF(nil, nil, programF, loaders...)
+
+	p := &TestPlan{
+		Options: TestUpdateOptions{HostF: hostF},
+	}
+
+	project := p.GetProject()
+
+	validateSnap := func(snap *deploy.Snapshot) {
+		assert.NotNil(t, snap)
+		assert.Nil(t, snap.VerifyIntegrity())
+		assert.Len(t, snap.Resources, 3)
+		assert.Equal(t, resource.URN("urn:pulumi:test::test::pkgA:m:typA::resA"), snap.Resources[1].URN)
+		assert.Equal(t, resource.URN("urn:pulumi:test::test::pkgA:m:typA::resB"), snap.Resources[2].URN)
+	}
+
+	// Run an update for initial state.
+	snap, err := TestOp(Update).Run(project, p.GetTarget(t, nil), p.Options, false, p.BackendClient, nil)
+	assert.NoError(t, err)
+	validateSnap(snap)
+
+	snap, err = TestOp(Destroy).Run(project, p.GetTarget(t, snap), TestUpdateOptions{
+		HostF: hostF,
+		UpdateOptions: UpdateOptions{
+			Targets: deploy.NewUrnTargets([]string{
+				"urn:pulumi:test::test::pkgA:m:typA::resA",
+			}),
+		},
+	}, false, p.BackendClient, nil)
+	assert.Error(t, err) // Expect error because we didn't specify the dependency as a target or TargetDependents
```

Comment:
Don't use `assert.Error` but `EqualError` or at least `ErrorContains` to check this is the error we expect.

Reply:
I can add it, but it's not actually very specific in this case: we get a generic error that says "step generator errored".

What's happening is we write out error messages to the user and then set `sawError = true`:

https://github.com/pulumi/pulumi/blob/3d37369cb4245c90262b172e962e09694fe70f1a/pkg/resource/deploy/step_generator.go#L1278-L1290

https://github.com/pulumi/pulumi/blob/3d37369cb4245c90262b172e962e09694fe70f1a/sdk/go/common/diag/errors.go#L79-L82

The actual `step generator errored` error we get here from `Run` is from this:

https://github.com/pulumi/pulumi/blob/3d37369cb4245c90262b172e962e09694fe70f1a/pkg/resource/deploy/deployment_executor.go#L326-L337

Reply:
Hmm ok, I might have a look at making that error better then. It would be good to be able to test in places like this which bit of the engine err'd even if we don't bubble that error value specifically back to the user. Was a big point of the bail rework we did. But no need to be done in this PR, can tidy that later.